File: 03/00_recursao/codes/puzzle/puzzle_print.c
#include <stdio.h>

int puzzle_print(int n,int tabs)
{
  int i;
  for(i=0;i<tabs;i++) printf("  ");
  printf("puzzle(%d)\n",n);
  if(n == 1) return(1);
  if(n % 2 == 0)
    return(puzzle_print(n/2,tabs+1));
  else return(puzzle_print(3*n+1,tabs+1));
}

int main(){
  int x;
  scanf("%d",&x);

  printf("%d\n",puzzle_print(x,0));
}

File: src/02/codes/recursao/puzzle_print.go
// "Puzzle" da Conjectura de Collatz, imprimindo a árvore de chamadas
// recursivas (indentada por profundidade).

package main

import (
	"fmt"
	"strings"
)

func puzzlePrint(n, tabs int) int {
	fmt.Printf("%spuzzle(%d)\n", strings.Repeat("  ", tabs), n)

	if n == 1 {
		return 1
	}
	if n%2 == 0 {
		return puzzlePrint(n/2, tabs+1)
	}
	return puzzlePrint(3*n+1, tabs+1)
}

func main() {
	var x int
	fmt.Scan(&x)

	fmt.Println(puzzlePrint(x, 0))
}
